### ISSUE TITLE
EndpointAutoConfiguration should be part of ManagementContextConfiguration

### ISSUE THREAD

**Issue by aantono** [spring-projects/spring-boot#7616]
Similar to `EndpointWebMvcManagementContextConfiguration`, it is probably best to have `EndpointAutoConfiguration` be also using `@ManagementContextConfiguration` instead of `@Configuration`.  This would allow one to define beans of type `HealthIndicator`, `PublicMetrics`, etc only in management context and have it be automatically picked up.

Currently if one declares those beans inside a class annotated with `@ManagementContextConfiguration`, while the beans get created, they don't get exposed via `/metrics` or `/health` endpoints, because the `MetricsEndpoint`, which autowires `PublicMetrics` beans, gets created inside main context and thus can't see any of the beans of that type created in the child management context.

**Comment by philwebb** [comment]
The management context is just one method that can be used to expose Endpoint beans (JMX being the most common other one). The `EndpointAutoConfiguration` class is for general endpoint configuration and isn't really tied to MVC concerns.

Is there any particular reason that you need your actual `Endpoint` beans to be in the management context?

**Comment by aantono** [comment]
The reason for having actual `Endpoint` bean to be int he management context is to pickup the custom `HealthIndicator`, `PublicMetrics`, etc beans that are also declared in the management context.  From the logical point of view, as I've explained in the description, it makes sense that those types of beans be part of the management context since they are not really part of the main application per say, they just provide management functions around it.  But if those are declared in the config that is annotated with `@ManagementContextConfiguration`, then they are not being auto-wired during the creation of `MetricsEndpoint`, as an example, so all the custom `PublicMetrics` beans are not exposed via the `/metrics` URL, same goes for the custom `HealthIndicators` and being able to see them in `/health`.  
Does this make sense?

**Comment by philwebb** [comment]
It's still not really clear to me _why_ you want to define beans in the management context. The management context is more of an internal concern designed to allow the management server to run on a different port.

The management context is also a child of the main context and I think it's preferable that beans like `HealthIndicator` and `PublicMetrics` are created in the child context so that everything has access to them.

Unless you're working on very specific actuator concerns, it's probably best to simply replace `@ManagementContextConfiguration` with plain `@Configuration` classes in your own code.

**Comment by aantono** [comment]
I guess I'm not fully following as to what is the purpose of the `@ManagementContextConfiguration` if you recommend to declare all the "management in nature" beans like `HealthIndicator` or `PublicMetrics` in the main context.  I was under the assumption that since those things are specifically for management, they should be logically placed in the proper context.

I know that the management context is a child of the main one and all the beans in the main are accessible from the child, but that does not go the other way, so since all the `*Endpoint` beans are created in the main context, they can't consume the parts defined in the management context.   So it sounds like you are recommending to create `HealthIndicator` and `PublicMetrics` beans in the main context instead of the management one. (I'm assuming you were meant to say that instead of what you actually typed: 

> The management context is also a child of the main context and I think it's preferable that beans like `HealthIndicator` and `PublicMetrics` are created in the **_child_** context so that everything has access to them.

)

**Comment by philwebb** [comment]
The management context setup is a little unusual. It probably wouldn't exist at all if it weren't for the fact that a different context is needed when running on a different port.

In a typical setup, where the management port is the same as the server port we simply put everything in the same context. The `EndpointWebMvcAutoConfiguration.EndpointWebMvcConfiguration` class uses `ManagementContextConfigurationsImportSelector` to load `@ManagementContextConfiguration` classes and they simply end up in the same context.

When the management port is different to the server port we try to keep the setup as similar as possible. We don't move the `HealthIndicator` or `Metrics` classes because it would just cause unnecessary complication in our configuration and introduce quite big differences in the way applications run.

Lets create some contrived example. Suppose a user wants to make use of a `HealthIndicator` in their own code. They could do something like this:

```java
@Service
public class MyService {
	
	private final DataSourceHealthIndicator indicator;

	public MyService(DataSourceHealthIndicator indicator) {
		this.indicator = indicator;
	}

	public void something() {
		Health health = this.indicator.health
		// Do something with health
	}

}
```

I know it's a totally stupid example, but it's quite conceivable that someone might expect that they can inject a `HealthIndicator` or `PublicMetrics` or whatever.

The problem we now have is that someone can write this code and it'll work fine if the management port is the same as the server port, but if they change the management port it suddenly stops working. This kind of unexpected change is really confusing for people so we'd rather keep things simple.

So, what we actually _really_ use `@ManagementContextConfiguration` for is to create a very small layer that starts the Servlet container on a different port and exposes the `Endpoints`. You can see this in `EndpointWebMvcChildContextConfiguration`. It's relatively minimal, it's job is to setup a `DispatcherServlet` etc and then import `@ManagementContextConfiguration` classes. We'll add a `EndpointHandlerMapping` and various MVC specific endpoints in the management context, but we leave the real `Endpoint` beans where they usually are (in the parent context).

Perhaps we should update the javadoc to make it clearer that it's a somewhat internal class and you probably won't need it in your own code.

To recap:

- `HealthIndicator`, `PublicMetrics` etc are beans that can be exposes in a number of different ways and for consistency we recommend keeping them in your main context.

- `@ManagementContextConfiguration` indicates that a configuration will be pulled into the management context if one exists (i.e. if you're run a different management port).

- You probably don't need to worry about `@ManagementContextConfiguration` in your own code, but we need to to make sure MVC concerns are configured in the correct context.
